File: tools/awali/awali/dyn/modules/determinize.cc
// This file is part of Awali.
// Copyright 2016-2023 Sylvain Lombardy, Victor Marsault, Jacques Sakarovitch
//
// Awali is a free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#ifndef DYN_DETERMINIZE_CC
#define DYN_DETERMINIZE_CC

#include <iostream>
#include <unordered_map>

#include <awali/dyn/loading/handler.hh>

#include <awali/dyn/modules/determinize.hh>
//This module must not be used with zmin or zmax...
//Only for lal automata

namespace awali {
  namespace dyn {

    automaton_t
    determinize (automaton_t aut, options_t opts)
    {
      if(aut->get_context()->weightset_name()=="B")
	return loading::call1<automaton_t>("determinize", "determinize", aut,
                                         opts[KEEP_HISTORY]);
      if(!opts[SAFE] || aut->get_context()->is_locally_finite_weightset())
        return loading::call1<automaton_t>("weighted_determinize", "determinize", aut);
      else
        throw std::runtime_error("Determinization is allowed only for locally finite weightsets");

    }


    automaton_t
    complement (automaton_t aut, options_t opts)
    {
      if (opts[IN_PLACE]) {
        loading::call1<void>("complement_here", "determinize", aut);
        return aut;
      }
      else
        return loading::call1<automaton_t>("complement", "determinize", aut);
    }


    automaton_t
    complete (automaton_t aut, options_t opts)
    {
      if (opts[IN_PLACE]) {
        loading::call1<void>("complete_here", "determinize", aut);
        return aut;
      }
      else
        return loading::call1<automaton_t>("complete", "determinize", aut);
    }


    automaton_t
    reduce (automaton_t aut)
    {
      return loading::call1<automaton_t>("reduce", "determinize", aut);
    }


    automaton_t
    left_reduce (automaton_t aut)
    {
      return loading::call1<automaton_t>("left_reduce", "determinize", aut);
    }


    bool
    is_deterministic (automaton_t aut)
    {
      return loading::call1<bool>("is_deterministic", "determinize", aut);
    }

    bool
    is_sequential (automaton_t aut)
    {
      return loading::call1<bool>("is_sequential", "determinize", aut);
    }

    bool
    is_complete (automaton_t aut)
    {
      return loading::call1<bool>("is_complete", "determinize", aut);
    }

    bool
    is_ambiguous (automaton_t aut)
    {
      return loading::call1<bool>("is_ambiguous", "determinize", aut);
    }

    automaton_t explore_by_length(automaton_t aut, unsigned depth) {
      return loading::call1<automaton_t>("explore_by_length", "determinize", aut, depth);
    }


    
    automaton_t explore_with_bound(automaton_t aut, weight_t bound) {
      std::string ws = aut->get_context()->weightset_name();
      context::weightset_description wd1 = context::weightset(ws);
      /* Although it is stated "N2", below, the test for type will work for any
       * bounded semiring */
      context::weightset_description wd2 = context::weightset("N2"); 
      if (ws == "N" || ws == "Z" || ws == "Noo" || wd1->type_ == wd2->type_)
        return loading::call1<automaton_t>("explore_with_bound", "determinize", aut, bound);
      throw std::runtime_error(
        "[dyn::explore_with_bound] It is required that the automaton is"
        " weighted over N, Z, Noo, or N<int>.");
    }
  }
}//end of ns awali::dyn

#endif

File: tools/awali/awalipy/examples/determinize.py
# This file is part of Awali.
# Copyright 2016-2023 Sylvain Lombardy, Victor Marsault, Jacques Sakarovitch
#
# Awali is a free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

import awali as vr
import sys

def key(set):
    #This is particularly inefficient.
    return str(sorted(set));




def determinize (A) :
    #B will be the DFA resulting from the power-set construction applied to A
    B = vr.Automaton(A.alphabet())

    #Maps a  (set of states of A)  to a  (state of B)
    set_to_state = dict()
# ================================================ Initial state computation ==
    init_state_of_B = B.add_state()
    B.set_initial(init_state_of_B)
    init_state_of_B_as_set = set()
    for p in A.initial_states() :
        init_state_of_B_as_set.add(p)
    set_to_state[key(init_state_of_B_as_set)]= init_state_of_B
# ================================================================ Main loop ==
    #The state-set of B as a list of (states of A)
    list_of_Bstates_as_sets = [init_state_of_B_as_set]
    for cur_Bstate_as_set in list_of_Bstates_as_sets:
        cur_Bstate = set_to_state[key(cur_Bstate_as_set)]
        for letter in A.alphabet() :
            new_Bstate_as_set = set()
            for cur_Astate in cur_Bstate_as_set :
                for q in A.successors(cur_Astate, letter) :
                    new_Bstate_as_set.add(q)

            #If the new state has never been reached before, it is added to B
            #and the appropriate variables are updated; otherwise the
            #corresponding state of B is computing using the map set_to_state.
            if new_Bstate_as_set not in list_of_Bstates_as_sets :
                new_Bstate = B.add_state()
                list_of_Bstates_as_sets.append(new_Bstate_as_set)
                set_to_state[key(new_Bstate_as_set)] = new_Bstate
            else :
                new_Bstate = set_to_state[key(new_Bstate_as_set)]

            B.set_transition(cur_Bstate, new_Bstate, letter)
# ================================================= Final states computation ==
    for cur_Bstate_as_set in list_of_Bstates_as_sets:
        for cur_Astate in cur_Bstate_as_set:
            if A.is_final(cur_Astate):
                B.set_final(set_to_state[key(cur_Bstate_as_set)])
                break;
    return B


class Determinizer(vr.Automaton):
    def __init__(self, source):
        vr.Automaton.__init__(self,source.alphabet())
        self.source = source
        self.states_as_sets = list()
        self.set_to_state = dict()
        self.states_to_visit_as_sets = list()
        self.remaining_letters = list()
        self.cur_state_as_set = None

        init_state_as_set = set()
        for q in self.source.states():
            if self.source.is_initial(q):
                init_state_as_set.add(q)
        self.states_as_sets.append(init_state_as_set)
        self.states_to_visit_as_sets.append(init_state_as_set)
        init_state = self.add_state(init_state_as_set)
        self.set_initial(init_state)
        self.set_to_state[key(init_state_as_set)] = init_state

    def add_state(self, new_state_as_set):
        name=''
        sep='{'
        for q in new_state_as_set:
            name+= sep+str(q)
            sep= ', '
        name+= '}'
        new_state= vr.Automaton.add_state(self,name)
        for q in new_state_as_set:
            if (self.source.is_final(q)):
                self.set_final(new_state)
                break;
        return new_state;

    def next(self):
        if (not self.has_work_left()):
            return
        if (len(self.remaining_letters) > 0):
            cur_letter= self.remaining_letters.pop()
        else:
            self.cur_state_as_set = self.states_to_visit_as_sets.pop()
            self.remaining_letters = list(self.alphabet())
            cur_letter= self.remaining_letters.pop(0)
        new_state_as_set = set()
        for cur_source_state in self.cur_state_as_set :
            for q in self.source.successors(cur_source_state, cur_letter) :
                new_state_as_set.add(q)
        if new_state_as_set not in self.states_as_sets:
            self.states_as_sets.append(new_state_as_set)
            new_state = self.add_state(new_state_as_set)
            self.states_to_visit_as_sets.append(new_state_as_set)
            self.set_to_state[key(new_state_as_set)] = new_state
        else:
            new_state = self.set_to_state[key(new_state_as_set)]
        cur_state = self.set_to_state[key(self.cur_state_as_set)]
        self.set_transition(cur_state, new_state, cur_letter)

    def has_work_left(self):
        return (len(self.states_to_visit_as_sets) + len(self.remaining_letters) > 0)

    def result(self):
        while (self.has_work_left()):
            self.next()
        return self;

if __name__ == '__main__':
    if (len(sys.argv) < 2) or (len(sys.argv) > 3) :
        print "Usage: determinize <in> [out]"
    else :
        A = vr.load(sys.argv[1])
        B = determinize(A)
        if (len(sys.argv) == 3) :
            B.save(sys.argv[2])
